core/src/telemetry/prometheus_server.rs

```diff
@@ -1,52 +1,47 @@
-use crate::telemetry::default_resource;
 use hyper::{
     header::CONTENT_TYPE,
     server::conn::AddrIncoming,
     service::{make_service_fn, service_fn},
     Body, Method, Request, Response, Server,
 };
-use opentelemetry::sdk::{
-    export::metrics::{aggregation::TemporalitySelector, AggregatorSelector},
-    metrics::{controllers, processors},
-};
-use opentelemetry_prometheus::{ExporterBuilder, PrometheusExporter};
-use prometheus::{Encoder, TextEncoder};
-use std::{collections::HashMap, convert::Infallible, net::SocketAddr, sync::Arc, time::Duration};
+use opentelemetry_prometheus::PrometheusExporter;
+use opentelemetry_sdk::metrics::reader::AggregationSelector;
+use prometheus::{Encoder, Registry, TextEncoder};
+use std::{convert::Infallible, net::SocketAddr};
 
 /// Exposes prometheus metrics for scraping
 pub(super) struct PromServer {
     bound_addr: AddrIncoming,
-    pub exporter: Arc<PrometheusExporter>,
+    registry: Registry,
 }
 
 impl PromServer {
     pub fn new(
         addr: SocketAddr,
-        aggregation: impl AggregatorSelector + Send + Sync + 'static,
-        temporality: impl TemporalitySelector + Send + Sync + 'static,
-        tags: &HashMap<String, String>,
-    ) -> Result<Self, anyhow::Error> {
-        let controller =
-            controllers::basic(processors::factory(aggregation, temporality).with_memory(true))
-                // Because Prom is pull-based, make this always refresh
-                .with_collect_period(Duration::from_secs(0))
-                .with_resource(default_resource(tags))
-                .build();
-        let exporter = ExporterBuilder::new(controller).try_init()?;
+        aggregation: impl AggregationSelector + Send + Sync + 'static,
+    ) -> Result<(Self, PrometheusExporter), anyhow::Error> {
+        let registry = Registry::new();
+        let exporter = opentelemetry_prometheus::exporter()
+            .with_aggregation_selector(aggregation)
+            .without_scope_info()
+            .with_registry(registry.clone());
         let bound_addr = AddrIncoming::bind(&addr)?;
-        Ok(Self {
-            exporter: Arc::new(exporter),
-            bound_addr,
-        })
+        Ok((
+            Self {
+                bound_addr,
+                registry,
+            },
+            exporter.build()?,
+        ))
     }
 
     pub async fn run(self) -> hyper::Result<()> {
         // Spin up hyper server to serve metrics for scraping. We use hyper since we already depend
         // on it via Tonic.
-        let expclone = self.exporter.clone();
+        let regclone = self.registry.clone();
```

Comment:
This clone isn't needed.

Reply:
It is - otherwise there's no way to move the registry into the closure w/o capturing `self`

Reply:
NM, you're right - forgot this fn is consuming self